### PR TITLE
Feat: Add plan steps

### PR DESCRIPTION
**Pull request by mykalmax**
The goal here is to emit events to UI on every state change for all stages during plan evaluation. We need to track when stage started when next change happed and when it ends. As well as explicitly set when model evaluation started and when it's finished. With all these we can show better statuses and updates on UI during each stage.

https://github.com/TobikoData/sqlmesh/assets/5644753/af25c376-2429-45f3-8243-dfea3fd2f334

### CODE REVIEW

File: web/client/src/library/components/plan/PlanApplyStageTracker.tsx [hunk 1]
```diff
@@ -0,0 +1,819 @@
...
+    () =>
+      Array.from(change_categorization.values()).reduce<
+        Record<string, boolean[]>
+      >((acc, { category, change }) => {
+        change?.indirect?.forEach(model => {
+          if (acc[model] == null) {
+            acc[model] = []
+          }
+
+          acc[model]?.push(category.value !== 1)
```
Comment (vchan) [inline web/client/src/library/components/plan/PlanApplyStageTracker.tsx:417]: What does the 1 (and 3 below) mean?

File: web/client/src/library/components/tasksOverview/TasksOverview.tsx [hunk 2]
```diff
@@ -172,69 +167,139 @@ function TasksDetails({
...
-                  </>
-                )}
-                {showVirtualUpdate && (
-                  <span className="inline-block whitespace-nowrap font-bold ml-2">
-                    Updated
-                  </span>
+                    <TaskDivider />
+                    {showProgress && (
+                      <>
+                        {task.end == null || task.start == null ? (
```
Comment (vchan) [inline web/client/src/library/components/tasksOverview/TasksOverview.tsx:209]: isNil?

File: web/client/src/library/components/tasksOverview/TasksOverview.tsx [hunk 3]
```diff
@@ -172,69 +167,139 @@ function TasksDetails({
...
+                    <TaskSize
+                      completed={task.completed}
+                      total={task.total}
+                      unit="batch"
+                    />
+                  )}
+                  <TaskDivider />
+                  {showProgress && (
+                    <>
+                      {task.end == null || task.start == null ? (
```
Comment (vchan) [inline web/client/src/library/components/tasksOverview/TasksOverview.tsx:273]: isNil?

File: web/client/src/library/pages/ide/IDE.tsx [hunk 4]
```diff
@@ -271,47 +284,61 @@ export default function PageIDE(): JSX.Element {
...
+
+    if (isTrue(meta?.done)) {
+      if (meta?.status === Status.success) {
+        console.log('Plan Apply Tracking Done Successfully')
+      } else if (meta?.status === Status.fail) {
+        console.log('Plan Apply Tracking Failed')
+      } else {
+        console.log('Plan Apply Tracking Done but status is unknown')
+      }
+    }
```
Comment (vchan) [inline web/client/src/library/pages/ide/IDE.tsx:306]: Is this for debugging?

File: web/client/src/models/initial.ts [hunk 5]
```diff
@@ -8,19 +8,25 @@ export class ModelInitial<T extends object = any> {
...
-      : new Proxy<InitialWithId<T>>(
-          Object.assign(initial ?? {}, {
-            id: uid(),
-          }),
-          {
-            set() {
-              throw new Error('Cannot change initial file')
+  constructor(initial?: Initial<T> | InitialWithId<T>) {
+    if (isNil(initial)) {
+      this._initial = Object.assign(initial ?? {}, {
```
Comment (vchan) [inline web/client/src/models/initial.ts:13]: Do you need the nullish coalescing operator if `initial` is already known to be null?

File: web/client/src/library/components/plan/PlanApplyStageTracker.tsx [hunk 6]
```diff
@@ -0,0 +1,819 @@
...
+}): JSX.Element {
+  const { change_categorization } = usePlan()
+
+  const categories = useMemo(
+    () =>
+      Array.from(change_categorization.values()).reduce<
+        Record<string, boolean[]>
+      >((acc, { category, change }) => {
+        change?.indirect?.forEach(model => {
+          if (acc[model] == null) {
```
[the remark this answers is not in the document]
Comment (mykalmax) [reply web/client/src/library/components/plan/PlanApplyStageTracker.tsx:413]: I started with `== null` but later added proper TS to hint that value after `isNil` `== null`. so now trying to switch to use `isNil`

File: web/client/src/library/components/plan/PlanApplyStageTracker.tsx (hunk 1 above)
Comment (mykalmax) [reply web/client/src/library/components/plan/PlanApplyStageTracker.tsx:417]: so openapi SnapshotChangeCategory converts to 
```
export const SnapshotChangeCategory = {
  NUMBER_1: 1, // BREAKING
  NUMBER_2: 2, // NON_BREAKING
  NUMBER_3: 3, // FORWARD_ONLY
  NUMBER_4: 4, // INDIRECT_BREAKING
  NUMBER_5: 5, // INDIRECT_NON_BREAKING
} as const
```

**Review by vchan — APPROVED** [review]

**Pull request merged**